src/main/java/com/google/api/gax/grpc/AbstractPage.java

```diff
@@ -29,76 +29,81 @@
  */
 package com.google.api.gax.grpc;
 
+import com.google.api.gax.core.Page;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.AbstractIterator;
 import com.google.common.collect.Iterables;
 import java.util.Iterator;
 
-public class PageContext<RequestT, ResponseT, ResourceT> {
+public abstract class AbstractPage<RequestT, ResponseT, ResourceT> implements Page<ResourceT> {
+
+  protected interface PageFactory<RequestT, ResponseT, ResourceT, PageT> {
+    PageT createPage(
+        UnaryCallable<RequestT, ResponseT> callable,
+        PagedListDescriptor<RequestT, ResponseT, ResourceT> pageDescriptor,
+        RequestT request,
+        CallContext context,
+        ResponseT response);
+  }
```

Comment:
I just thought of an idea - let me know if this is unworkable. The idea is to avoid the PageFactory abstraction by instead using the idea of "page templating" - create an "empty page" as a template for the first real page (from within the client code); create following pages using the last page as a "template" for the next page. I think this would require less generated code and less parameters passed around. Let me know what you think.